augur/subsample.py

```diff
@@ -0,0 +1,366 @@
+"""
+Subsample sequences from an input dataset.
+
+The input dataset can consist of a metadata file, a sequences file, or both.
+
+See documentation page for details on configuration.
+"""
+
+import argparse
+import os
+import subprocess
+import tempfile
+import yaml
+from concurrent.futures import ThreadPoolExecutor, as_completed
+from typing import Any, Dict, List, Optional, Tuple, Union
+from augur import filter as augur_filter
+from augur.argparse_ import ExtendOverwriteDefault, SKIP_AUTO_DEFAULT_IN_HELP
+from augur.errors import AugurError
+from augur.io.metadata import DEFAULT_DELIMITERS, DEFAULT_ID_COLUMNS
+from augur.io.print import print_err
+from augur.utils import augur
+from augur.validate import load_json_schema, validate_json
+
+
+BooleanFlags = Tuple[str, Optional[str]]
+"""
+Type for a boolean pair of augur filter command line flags that configure the
+same option. When there is no second flag, absence of the first flag indicates
+the default behavior.
+"""
+
+AugurFilterOption = Union[str, BooleanFlags]
+"""
+Type for an augur filter command line option. Either a single option or boolean
+pair of flags.
+"""
+
+FilterArgs = Dict[str, Any]
+"""
+Augur filter arguments stored as a mapping from option to value.
+"""
+
+
+GLOBAL_CLI_OPTIONS: Dict[str, AugurFilterOption] = {
+    "metadata": "--metadata",
+    "metadata_chunk_size": "--metadata-chunk-size",
+    "metadata_delimiters": "--metadata-delimiters",
+    "metadata_id_columns": "--metadata-id-columns",
+    "sequences": "--sequences",
+    "sequence_index": "--sequence-index",
+    "seed": "--subsample-seed",
+}
+"""
+Mapping of argparse namespace variable name to augur filter option.
+These are sent to both intermediate and final augur filter calls.
+"""
+
+
+FINAL_CLI_OPTIONS: Dict[str, AugurFilterOption] = {
+    "output_metadata": "--output-metadata",
+    "output_sequences": "--output-sequences",
+    "skip_checks": ("--skip-checks", None),
+}
+"""
+Mapping of argparse namespace variable name to augur filter option.
+These are sent to only the final augur filter call.
+"""
+
+
+# NOTE: If you edit any of these values, please re-run
+# devel/regenerate-subsample-schema and commit the schema changes too.
+SAMPLE_CONFIG: Dict[str, AugurFilterOption] = {
+    "exclude": "--exclude",
+    "exclude_all": ("--exclude-all", None),
+    "exclude_ambiguous_dates_by": "--exclude-ambiguous-dates-by",
+    "exclude_where": "--exclude-where",
+    "include": "--include",
+    "include_where": "--include-where",
```

Comment:
Another renaming proposal before initial release:

- `exclude` → `exclude_files` or `exclude_id_files`
- `include` → `include_files` or `include_id_files`

The current names are ambiguous because we already have multiple `exclude_*` and `include_*` options. This ambiguity is tolerable for `augur filter` command line options since confused users can quickly check `augur filter --help`. However, in `augur subsample` config, the YAML keys should be more self-descriptive, as users won’t have the same immediate context or help reference.

Reply:
`exclude`/`include` is used pretty widely across pathogen Snakemake configs that I don't think it's too ambiguous for current users, but I can see how it might be ambiguous for new users. However, I'd expect new users to reference the docs page to fully understand each config param rather than infer their usage based on the param name. 

I don't feel too strongly either way, but if renaming, then I'd prefer `exclude_id_files`/ `include_id_files`.